src/lib/autoinstall/clients/inst_autoinit.rb

```diff
@@ -0,0 +1,368 @@
+              "the AutoYaST profile or use full installation medium.")
+            Yast::Popup.LongError(msg) # No timeout because we are stopping the installation/upgrade.
+
+            return :abort
+          end
+
+          suse_register if !Mode.autoupgrade
+        # offline registration need here to init software management according to picked product
+        # or autoupgrade without scc
+        elsif Y2Packager::MediumType.offline?
+          product = Yast::AutoinstFunctions.selected_product
+
+          # if addons contain relurl. If so, then product have to defined, otherwise relurl
+          # cannot be expanded in autoupgrade
+          addon_profile = (Yast::Profile.current["add-on"] || {})
+          addons = (addon_profile["add_on_products"] || []) + (addon_profile["add_on_others"] || [])
+          addons_relurl = addons.find { |a| (a["relurl"] || "") =~ /relurl:\/\// }
+
+          # duplicite code, but for offline medium we need to do it before system is initialized as
+          # we get need product for initialize libzypp, but for others we need first init libzypp
+          # to get available products.
+          if product
+            show_popup = true
+            base_url = Yast::InstURL.installInf2Url("")
+            log_url = Yast::URL.HidePassword(base_url)
+            Yast::Packages.Initialize_StageInitial(show_popup, base_url, log_url, product.dir)
+            # select the product to install
+            Yast::Pkg.ResolvableInstall(product.details.product, :product, "")
+            # initialize addons and the workflow manager
+            Yast::AddOnProduct.SetBaseProductURL(base_url)
+            Yast::WorkflowManager.SetBaseWorkflow(false)
+            Yast::AutoinstFunctions.reset_product
+          # report error only for installation or if autoupgrade contain addon with relative url.
+          # This way autoupgrade for Full medium on registered system
+          # can autoupgrade with empty profile.
+          elsif !Yast::Mode.autoupgrade || addons_relurl
+            if Yast::Mode.autoupgrade && addons_relurl
+              msg = _("None or wrong base product has been defined " \
+                "in the AutoYaST configuration file. " \
+                "It needs to be specified as base for addons that use relurl scheme." \
+                "Please check the <b>products</b> entry in the <b>software</b> section.<br><br>" \
+                "Following base products are available:<br>")
+            else
+              msg = _("None or wrong base product has been defined " \
+                "in the AutoYaST configuration file. " \
+                "Please check the <b>products</b> entry in the <b>software</b> section.<br><br>" \
+                "Following base products are available:<br>")
+            end
+            Yast::AutoinstFunctions.available_base_products.each do |product|
+              # FIXME: here we abuse knowledge that base product is ProductLocation and not Product
```

Comment:
I do not understand the comment :)

Reply:
well, we have Product and ProductLocation. It looks quite similar, but few details differ and here we use some details that is only in ProductLocation and not in Product. Not sure how to write better. But in general it is a bit hacky.